Tests/NIOTests/CodecTest.swift

```diff
@@ -237,6 +237,95 @@ public class ByteToMessageDecoderTest: XCTestCase {
         XCTAssertEqual(decoder.cumulationBuffer!.readableBytes, 3072)
         XCTAssertEqual(decoder.cumulationBuffer!.readerIndex, 0)
     }
+
+    func testDecoderReentranceChannelRead() throws {
+        let channel = EmbeddedChannel()
+        defer {
+            XCTAssertNoThrow(try channel.finish())
+        }
+
+        class TestDecoder: ByteToMessageDecoder {
+
+            typealias InboundIn = ByteBuffer
+            typealias InboundOut = ByteBuffer
+
+            var cumulationBuffer: ByteBuffer?
+            var triggeredReentrace: Bool = false
+            var decode: Bool = false
+
+            func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) throws -> DecodingState {
+                guard !self.decode else {
+                    XCTAssertTrue(self.triggeredReentrace)
+                    return .needMoreData
+                }
+                self.decode = true
+                defer {
+                    self.decode = false
+                }
+
+                if !self.triggeredReentrace {
+                    self.triggeredReentrace = true
```

Comment:
Rather than set this here, it might be better to let the re-entrant call set it (where you currently have an `XCTAssertTrue`) and then check it at the end of the test. Otherwise this test could pass if you never re-entered at all.